File: task/src/task/cmd.nim
import std/[options, times]
import filter
import types

proc list*(tasks: seq[Task], filter: string, includeFutureOffset: int, limit: int): seq[Task] =
  for i, task in parseFilter(filter).applyFilter(tasks):
    if limit >= 0 and i > limit:
      break

    if includeFutureOffset >= 0 and task.details.next.isSome():
      if task.details.next.get() > now() + includeFutureOffset.days:
        continue

    result.add(task)

proc listWithNext*(tasks: seq[Task], filter: string, includeFutureOffset: int, limit: int): seq[Task] =
  var i = 0
  for task in list(tasks, filter, includeFutureOffset, limit):
    if task.details.next.isSome():
      i += 1
      if limit >= 0 and i > limit:
        break

      result.add(task)

File: task/task/cmd.py
import datetime
import functools
import itertools
from collections.abc import Iterable
from collections.abc import Iterator

from .schema import Filter
from .schema import Task


def filter_max_ahead(task: Task, days: int) -> bool:
    if days >= 0 and task.details:
        target = datetime.date.today() + datetime.timedelta(days=days)
        if task.details.next_ > target:
            return False

    return True


def load(
    tasks: Iterable[Task], filter_: str, days: int, limit: int,
) -> Iterator[Task]:
    tasks = functools.reduce(Filter.apply, Filter.parse(filter_), tasks)
    tasks = (x for x in tasks if filter_max_ahead(x, days))
    if limit >= 0:
        tasks = itertools.islice(tasks, limit)

    yield from tasks


def load_with_next(
    tasks: Iterable[Task], filter_: str, days: int, limit: int,
) -> Iterator[Task]:
    tasks = load(tasks, filter_, days, -1)
    tasks = (x for x in tasks if x.details)
    if limit >= 0:
        tasks = itertools.islice(tasks, limit)

    yield from tasks
